eth/executionclient/managed_client.go

```diff
@@ -0,0 +1,192 @@
+package executionclient
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/ethclient"
+	"go.uber.org/zap"
+)
+
+// ManagedClient manages an individual Ethereum client connection.
+type ManagedClient struct {
+	addr                        string
+	client                      *ethclient.Client
+	mu                          sync.RWMutex
+	healthy                     bool
+	logger                      *zap.Logger
+	reconnectionInitialInterval time.Duration
+	reconnectionMaxInterval     time.Duration
+	syncDistanceTolerance       uint64
+	quit                        chan struct{}
+	wg                          sync.WaitGroup
+}
+
+// NewManagedClient creates and starts a new ManagedClient.
+func NewManagedClient(ctx context.Context, addr string, logger *zap.Logger, initialInterval, maxInterval time.Duration, syncDistanceTolerance uint64) (*ManagedClient, error) {
+	mc := &ManagedClient{
+		addr:                        addr,
+		logger:                      logger,
+		reconnectionInitialInterval: initialInterval,
+		reconnectionMaxInterval:     maxInterval,
+		syncDistanceTolerance:       syncDistanceTolerance,
+		quit:                        make(chan struct{}),
+	}
+
+	err := mc.connect(ctx)
+	if err != nil {
+		mc.logger.Error("Initial connection failed", zap.String("address", addr), zap.Error(err))
+		return nil, fmt.Errorf("connect: %w", err)
+	}
+
+	mc.wg.Add(1)
+	go mc.monitorHealth()
+
+	return mc, nil
+}
+
+// connect establishes a connection to the Ethereum client.
+func (mc *ManagedClient) connect(ctx context.Context) error {
```

Comment:
if this happens only once at startup maybe the mutex protection (for the client) is not needed

Reply:
in `attemptReconnect` too

Reply:
I just changed the implementation to use multi-client, so I'm resolving this